### DISCUSSION TITLE
Custom PyTree Backend -> Equinox Backend

### DISCUSSION THREAD

**Discussion by jejjohnson** [ucl-bug/jaxdf#110] [Ideas]
Hi, 

I really like this library a lot! It is the cleanest API for PDEs that I have seen that's intuitive and fairly extendible! I am working with PDEs for oceanography applications like inverse problems and surrogates and I would like to use this API to make some PDEs for Quasi-Geostrophic equations and Shallow Water equations. 

I was walking through the library bit by bit to understand how you all built this to see if it would work for my use-case. There are 2 things I wanted to ask if you had an example:
* I saw that you have listed in the new update that you have removed the time-steppers and you suggested using `diffrax`. It wasn't [updated in the docs](https://ucl-bug.github.io/jaxdf/notebooks/quickstart/#euler-integration) but I saw the latest release and code-base. Do you have an example of how to use this with diffrax? I think training the network parameters should be straight forward. However, I'm not sure how it works with mixing and matching different definitions of `pytrees` for the pytrees defined in `equinox` and the pytrees define in this library. (But perhaps I am wrong about this). 
* Related to the above question, for the [PINNS](https://ucl-bug.github.io/jaxdf/notebooks/pinn_burgers/) and [custom CNN](https://ucl-bug.github.io/jaxdf/notebooks/quickstart/#more-complex-operators) example, I see you used a CNN but you defined the parameters from scratch. That can get extremely cumbersome after a while so I wanted to know if it was compatible with other libraries like `equinox` or `flax`. If you have some examples already, that would be great to see! (Again, perhaps it's easier than I think it is...)

---

Also in relation to those two questions, the crux thing I wanted to ask was if you had any intention of converting/upgrading the current framework to use an already implemented library for `pytrees`, e.g. `equinox`? For me it seems like a good choice especially to couple this with the already "complete" ODE library `diffrax`. [In fact, a new example was just posted for how one could solve PDEs using `diffrax`](https://docs.kidger.site/diffrax/examples/nonlinear_heat_pde/). But I wanted to know the motivations and future directions for the package. I suppose there are 3 potential options below:

1) Integrate `equinox` into this `jaxdf` library as the back-end to make everything compatible with the `diffrax` library. I would definitely love to help with this. The downside of course is that now it is no longer agnostic and it might be difficult to mix and match libraries.

2) with your permission of course, I could also do a small rewrite of the library using the `equinox` backend for the pytrees. I could rename the library `PDiffrax` or `eqx-df`. I personally wouldn't mind doing it because it is good for my use case and it would be an excellent exercise for me to become more familiar with JAX.

3) If there is a clever way to "transform" the pytree classes here to other frameworks, e.g. `equinox`, `flax`, then that would probably be the best case! (perhaps you already have examples of this?)

**Comment by astanziola** [comment]
Hello @jejjohnson ! Very sorry for missing this, but I'm taking some time away from academia so I was not very active on the repo. I believe that integrating equinox (and diffrax!) is a great idea, do you mind if we have a more detailed conversation when I'm back on this? 

In general, however, this project is LGPL exactly because I'm more than happy for people to contribute to it or make their own versions / forks. 
However, having jaxdf work with diffrax and equinox, or generally having custom backends for pytrees as you mentioned, is something that would be extremely useful. Let me ping you in a few weeks, if that works for you :)

**Comment by astanziola** [comment]
Also, regarding the possibility of transforming classes to flax: you can always transform a `jaxdf` function into a pure jax one by wrapping / unwrapping fields, for example

```python
domain = Domain(...)

def func(x: jnp.ndarray):
  u = FourierSeries(params = x, domain=domain)
  b = some_jaxdf_operator(u)
  return b.on_grid
```

This should work with every `jax`-based framerowk. 

For example, [in this class](https://github.com/ucl-bug/lbs/blob/6203890e73394b981730504c788f0c6c152dc732/bno/modules.py#L89) we are using `jaxdf` operators and fields inside a `Flax` module.

**Comment by jejjohnson** [comment]
Thank you for your above comments! I would love to link up sometime in June if you have some time. 

I've been working silently on trying to improve my understanding of PDEs and how one can use Jax. I did a pure functional approach with [equinox](https://github.com/patrick-kidger/equinox/tree/main) and explicit finite difference schemes using [finitediffX](https://github.com/ASEM000/FiniteDiffX/tree/main) (and [kernex](https://github.com/ASEM000/Kernex)) and timesteppers using [diffrax](https://github.com/patrick-kidger/diffrax), e.g. see my attempt at the [12 steps to Navier-Stokes](https://github.com/jejjohnson/jaxsw/tree/eman/notebooks/dev/eman/12_steps_2_NS). I have to say the combination of libraries made working with PDEs a breeze. My ultimate goal is to work with [simple ocean-like PDEs](https://github.com/jejjohnson/jaxsw/blob/eman/notebooks/dev/eman/12_steps_2_NS/5.3_QG.ipynb) to work with some parameterisations but I think everything I've learned along the way could be very useful for anyone working with PDEs.

I would like to advertise this a bit to my numerical PDE peers so that they can see how we can jax-ify all the things. Of course the next step is the jaxdf implicit/automatic way of doing the discretisation. So, again, I would love to link up with you and perhaps discuss ways of doing this. I think the way jaxdf does it is the most intuitive I've seen and it also mirrors a [recent example in the diffrax framework](https://docs.kidger.site/diffrax/examples/nonlinear_heat_pde/) for the heat equation.

**Comment by astanziola** [comment]
Hi! I'm finally back full time on this, sorry for the slow response time so far. Shall we have a chat about this in more details? I made a discord server for this project (you can see it in the README) or I can send you an email and organize a call.

**Reply by jejjohnson** [reply]
Hi, I have joined the discord server. We can talk and setup a meeting time there! Talk soon!
